File: python/nothing/sideload.py
"""APK / split-APK sideload helpers."""

from pathlib import Path

from .device import run
from .exceptions import AdbError
from .models import DeviceInfo


def action_sideload(device: DeviceInfo, apk_path: str, downgrade: bool = False) -> None:
    """
    Install an APK or a directory of split APKs.

    - Single file (.apk)    → adb install -r [-d]
    - Directory             → adb install-multiple -r [-d] *.apk
    - downgrade=True        → adds -d flag (allows lower versionCode)
    """
    path = Path(apk_path)
    if not path.exists():
        raise AdbError(f"Path not found: {path}")

    if path.is_dir():
        apks = sorted(path.glob("*.apk"))
        if not apks:
            raise AdbError(f"No .apk files found in {path}")
        print(f"\nSplit APK install ({len(apks)} parts):")
        for a in apks:
            print(f"  {a.name}")
        cmd = ["adb", "-s", device.serial, "install-multiple", "-r"]
        if downgrade:
            cmd.append("-d")
        cmd.extend(str(a) for a in apks)
    else:
        if path.suffix.lower() != ".apk":
            raise AdbError(f"Expected a .apk file or directory, got: {path.name}")
        print(f"\nInstalling {path.name} ({path.stat().st_size // 1024} KB)...")
        cmd = ["adb", "-s", device.serial, "install", "-r"]
        if downgrade:
            cmd.append("-d")
        cmd.append(str(path))

    r = run(cmd, capture=True)
    output = (r.stdout + r.stderr).strip()

    if r.returncode == 0 and "Success" in output:
        print(f"[OK] Installed successfully.")
    elif "INSTALL_FAILED_VERSION_DOWNGRADE" in output:
        raise AdbError(
            "Install blocked: target version is older than installed.\n"
            "Use --downgrade to allow lower versionCode installs."
        )
    elif "INSTALL_FAILED_ALREADY_EXISTS" in output:
        raise AdbError("Package already installed. Use -r flag (already included) — "
                       "if still failing, uninstall first.")
    else:
        raise AdbError(f"Install failed: {output}")

File: go/internal/sideload/sideload.go
// Package sideload provides APK / split-APK sideload helpers.
package sideload

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

// ActionSideload installs a single APK or a directory of split APKs.
func ActionSideload(serial, apkPath string, downgrade bool) error {
	info, err := os.Stat(apkPath)
	if err != nil {
		return nterrors.AdbError(fmt.Sprintf("Path not found: %s", apkPath))
	}

	var cmd []string

	if info.IsDir() {
		apks, err := filepath.Glob(filepath.Join(apkPath, "*.apk"))
		if err != nil || len(apks) == 0 {
			return nterrors.AdbError(fmt.Sprintf("No .apk files found in %s", apkPath))
		}
		fmt.Printf("\nSplit APK install (%d parts):\n", len(apks))
		for _, a := range apks {
			fmt.Printf("  %s\n", filepath.Base(a))
		}
		cmd = []string{"adb", "-s", serial, "install-multiple", "-r"}
		if downgrade {
			cmd = append(cmd, "-d")
		}
		cmd = append(cmd, apks...)
	} else {
		ext := strings.ToLower(filepath.Ext(apkPath))
		if ext != ".apk" {
			return nterrors.AdbError(fmt.Sprintf("Expected a .apk file or directory, got: %s", filepath.Base(apkPath)))
		}
		fmt.Printf("\nInstalling %s (%d KB)...\n", filepath.Base(apkPath), info.Size()/1024)
		cmd = []string{"adb", "-s", serial, "install", "-r"}
		if downgrade {
			cmd = append(cmd, "-d")
		}
		cmd = append(cmd, apkPath)
	}

	stdout, stderr, code := adb.Run(cmd)
	output := strings.TrimSpace(stdout + stderr)

	if code == 0 && strings.Contains(output, "Success") {
		fmt.Println("[OK] Installed successfully.")
		return nil
	}
	if strings.Contains(output, "INSTALL_FAILED_VERSION_DOWNGRADE") {
		return nterrors.AdbError("Install blocked: target version is older than installed.\nUse --downgrade to allow lower versionCode installs.")
	}
	if strings.Contains(output, "INSTALL_FAILED_ALREADY_EXISTS") {
		return nterrors.AdbError("Package already installed. Use -r flag (already included) — if still failing, uninstall first.")
	}
	return nterrors.AdbError(fmt.Sprintf("Install failed: %s", output))
}
